File: day3/main.rs
use std::fs;
use std::time::Instant;

fn best_two(line: &str) -> i64 {
    let bytes = line.trim().as_bytes();
    if bytes.len() < 2 {
        return 0;
    }
    let n = bytes.len();
    let mut suffix = vec![b'0'; n + 1];
    for i in (0..n).rev() {
        suffix[i] = suffix[i + 1].max(bytes[i]);
    }
    let mut best: i64 = -1;
    for i in 0..n - 1 {
        let val = ((bytes[i] - b'0') as i64) * 10 + (suffix[i + 1] - b'0') as i64;
        if val > best {
            best = val;
        }
    }
    best
}

fn best_k(line: &str, k: usize) -> i64 {
    let mut digits: Vec<u8> = line.trim().bytes().collect();
    let mut drop = digits.len().saturating_sub(k);
    let mut stack: Vec<u8> = Vec::with_capacity(k);
    for d in digits.drain(..) {
        while drop > 0 && !stack.is_empty() && *stack.last().unwrap() < d {
            stack.pop();
            drop -= 1;
        }
        stack.push(d);
    }
    stack.truncate(k);
    stack.iter().fold(0i64, |acc, &d| acc * 10 + (d - b'0') as i64)
}

fn main() {
    let path = if fs::metadata("input_eric.txt").is_ok() {
        "input_eric.txt"
    } else {
        "input.txt"
    };
    let content = fs::read_to_string(path).expect("input");
    let lines: Vec<&str> = content.lines().collect();

    let start = Instant::now();
    let mut p1: i64 = 0;
    let mut p2: i64 = 0;
    for line in lines {
        p1 += best_two(line);
        p2 += best_k(line, 12);
    }
    let elapsed = start.elapsed().as_secs_f64() * 1000.0;
    println!(
        "max-2-digit-sum={} max-12-digit-sum={} elapsed_ms={:.3}",
        p1, p2, elapsed
    );
}

File: day3/main.ts
import { readFileSync, existsSync } from "fs";
import { performance } from "perf_hooks";

function bestTwo(line: string): bigint {
  const d = line.trim();
  const n = d.length;
  if (n < 2) return 0n;
  const suffix: number[] = new Array(n + 1).fill(0);
  for (let i = n - 1; i >= 0; i--) {
    suffix[i] = Math.max(suffix[i + 1], d.charCodeAt(i));
  }
  let best = -1n;
  for (let i = 0; i < n - 1; i++) {
    const val =
      BigInt(d.charCodeAt(i) - 48) * 10n + BigInt(suffix[i + 1] - 48);
    if (val > best) best = val;
  }
  return best;
}

function bestK(line: string, k: number): bigint {
  const digits = line.trim().split("").map((c) => c.charCodeAt(0));
  let drop = digits.length - k;
  const stack: number[] = [];
  for (const d of digits) {
    while (drop > 0 && stack.length && stack[stack.length - 1] < d) {
      stack.pop();
      drop--;
    }
    stack.push(d);
  }
  const kept = stack.slice(0, k);
  return kept.reduce((acc, d) => acc * 10n + BigInt(d - 48), 0n);
}

const inputFile = existsSync("input_eric.txt") ? "input_eric.txt" : "input.txt";
const t0 = performance.now();
const lines = readFileSync(inputFile, "utf8").trim().split(/\r?\n/);
let p1 = 0n;
let p2 = 0n;
for (const line of lines) {
  p1 += bestTwo(line);
  p2 += bestK(line, 12);
}
const elapsed = performance.now() - t0;
console.log(
  `max-2-digit-sum=${p1} max-12-digit-sum=${p2} elapsed_ms=${elapsed.toFixed(
    3,
  )}`,
);
